pkg/initpkg/context.go

```diff
@@ -0,0 +1,277 @@
+func WithLicense(license string) Option {
+	return func(ctx *Context) error {
+		ctx.License = license
+		return nil
+	}
+}
+
+// WithLayout overrides the Melange build strategy selected for
+// a package.
+func WithLayout(layout string) Option {
+	return func(ctx *Context) error {
+		ctx.Layout = layout
+		return nil
+	}
+}
+
+// New creates a new Context, which is used with Run.
+func New(opts ...Option) (*Context, error) {
+	ctx := Context{}
+
+	for _, opt := range opts {
+		if err := opt(&ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	if ctx.URI == "" {
+		return nil, fmt.Errorf("uri is not set")
+	}
+
+	return &ctx, nil
+}
+
+// InterpretGithub fills in package information for a GitHub project, and
+// fetches the source, then synthesizes a git-checkout node on the Melange
+// pipeline.
+func (ctx *Context) InterpretGitHub(u *url.URL) error {
+	rawURI := u.String()
+	re := regexp.MustCompile(`https://github\.com/([^/]+)/([^/]+)`)
+
+	match := re.FindStringSubmatch(rawURI)
+	if len(match) != 3 {
+		return fmt.Errorf("github project URI is malformed")
+	}
+
+	log.Printf("github repo owner: %s", match[1])
+	log.Printf("package name: %s", match[2])
+
+	return nil
+}
+
+// InterpretTarball fills in package information for a tarball, and fetches it,
+// synthesizing a fetch node on the Melange pipeline.
+func (ctx *Context) InterpretTarball(u *url.URL) error {
+	rawURI := u.String()
+	pattern := `([^\/]+)-([\d\.]+)\.(tar\.(gz|xz|bz2|lz|Z)|tgz|tbz2|txz|tlz)`
+
+	re := regexp.MustCompile(pattern)
+	if re == nil {
+		return fmt.Errorf("unable to compile version regex")
+	}
+
+	match := re.FindStringSubmatch(rawURI)
+	if len(match) < 3 {
+		return fmt.Errorf("unable to deduce package identity from URL")
+	}
+
+	ctx.Build.Package.Name = match[1]
+	ctx.Build.Package.Version = match[2]
+
+	downloadedFile, err := melange_util.DownloadFile(rawURI)
+	if err != nil {
+		return fmt.Errorf("downloading sources: %w", err)
+	}
+	defer os.Remove(downloadedFile)
+
+	expectedSHA512, err := melange_util.HashFile(downloadedFile, sha512.New())
+	if err != nil {
+		return fmt.Errorf("hashing sources: %w", err)
+	}
+
+	log.Printf("fetched source, hashed: %s", expectedSHA512)
+
+	pipelineNode := melange_build.Pipeline{
+		Uses: "fetch",
+		With: map[string]string{
+			"uri":             strings.ReplaceAll(rawURI, match[2], "${{package.version}}"),
+			"expected-sha512": expectedSHA512,
+		},
+	}
+	ctx.Build.Pipeline = append(ctx.Build.Pipeline, pipelineNode)
+
+	unpacker, err := NewUnpacker(ctx)
+	if err != nil {
+		return fmt.Errorf("initializing unpacker: %w", err)
+	}
+
+	if err := unpacker.Unpack(downloadedFile); err != nil {
+		return fmt.Errorf("unpacking sources: %w", err)
+	}
+
+	log.Printf("unpacked sources to %s", ctx.WorkDir)
+
+	return nil
+}
+
+// InterpretURI interprets the provided source code URI and fills in project
+// information as well as pipeline instructions in how to fetch the sources.
+func (ctx *Context) InterpretURI(uri string) error {
```

Comment:
I would not use `ctx` as a receiver because it collides with the idiomatic name for`context.Context` parameters.

(I'd also rename the struct from `Context` to something else so we don't have a dozen different flavors of Context floating around, but that's less important than the receiver name.)